From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Paul Praschl <praschlpaul@g-p.at>
Date: Mon, 27 Jul 2026 10:22:52 +0200
Subject: [PATCH 057/117] wifi: brcmfmac: propagate P2P firmware setup errors

P2P firmware preparation consists of required down, APSTA, up,
discovery-disable and address-override operations. Most errors and the
helper return value were previously ignored. Propagate each failure
immediately so interface creation does not continue into later commands
and event waits after setup has already failed.

If a successful BRCMF_C_DOWN is followed by an APSTA or BRCMF_C_UP
failure, make one best-effort BRCMF_C_UP attempt before returning.
Preserve the original setup error and report a separate restoration
failure without changing the normal WLAN client path.

Signed-off-by: Paul Praschl <praschlpaul@g-p.at>

Orchard-replay-of: 5ec90a4789a2d713cb7c38c7ac1f5af97a3fe549
---
 .../broadcom/brcm80211/brcmfmac/p2p.c         | 42 +++++++++++++++----
 1 file changed, 33 insertions(+), 9 deletions(-)

diff --git a/drivers/net/wireless/broadcom/brcm80211/brcmfmac/p2p.c b/drivers/net/wireless/broadcom/brcm80211/brcmfmac/p2p.c
index 813cc15d8..784cf1a58 100644
--- a/drivers/net/wireless/broadcom/brcm80211/brcmfmac/p2p.c
+++ b/drivers/net/wireless/broadcom/brcm80211/brcmfmac/p2p.c
@@ -455,24 +455,47 @@ static void brcmf_p2p_print_actframe(bool tx, void *frame, u32 frame_len)
 static int brcmf_p2p_set_firmware(struct brcmf_if *ifp, u8 *p2p_mac)
 {
 	struct brcmf_pub *drvr = ifp->drvr;
-	s32 ret = 0;
+	bool fw_down = false;
+	s32 restore_ret;
+	s32 ret;
+
+	ret = brcmf_fil_cmd_int_set(ifp, BRCMF_C_DOWN, 1);
+	if (ret)
+		goto fail;
+	fw_down = true;
+
+	ret = brcmf_fil_iovar_int_set(ifp, "apsta", 1);
+	if (ret)
+		goto fail;
 
-	brcmf_fil_cmd_int_set(ifp, BRCMF_C_DOWN, 1);
-	brcmf_fil_iovar_int_set(ifp, "apsta", 1);
-	brcmf_fil_cmd_int_set(ifp, BRCMF_C_UP, 1);
+	ret = brcmf_fil_cmd_int_set(ifp, BRCMF_C_UP, 1);
+	if (ret)
+		goto fail;
+	fw_down = false;
 
 	/* In case of COB type, firmware has default mac address
 	 * After Initializing firmware, we have to set current mac address to
 	 * firmware for P2P device address. This must be done with discovery
 	 * disabled.
 	 */
-	brcmf_fil_iovar_int_set(ifp, "p2p_disc", 0);
+	ret = brcmf_fil_iovar_int_set(ifp, "p2p_disc", 0);
+	if (ret)
+		goto fail;
 
 	ret = brcmf_fil_iovar_data_set(ifp, "p2p_da_override", p2p_mac,
 				       ETH_ALEN);
-	if (ret)
-		bphy_err(drvr, "failed to update device address ret %d\n", ret);
+	if (!ret)
+		return 0;
+
+fail:
+	if (fw_down) {
+		restore_ret = brcmf_fil_cmd_int_set(ifp, BRCMF_C_UP, 1);
+		if (restore_ret)
+			bphy_err(drvr, "failed to restore firmware ret %d\n",
+				 restore_ret);
+	}
 
+	bphy_err(drvr, "failed to prepare P2P firmware ret %d\n", ret);
 	return ret;
 }
 
@@ -2200,7 +2223,9 @@ static struct wireless_dev *brcmf_p2p_create_p2pdev(struct brcmf_p2p_info *p2p,
 	}
 
 	brcmf_p2p_generate_bss_mac(p2p, addr);
-	brcmf_p2p_set_firmware(pri_ifp, p2p->dev_addr);
+	err = brcmf_p2p_set_firmware(pri_ifp, p2p->dev_addr);
+	if (err)
+		goto fail;
 
 	brcmf_cfg80211_arm_vif_event(p2p->cfg, p2p_vif);
 	brcmf_fweh_p2pdev_setup(pri_ifp, true);
@@ -2567,4 +2592,3 @@ void brcmf_p2p_detach(struct brcmf_p2p_info *p2p)
 	/* just set it all to zero */
 	memset(p2p, 0, sizeof(*p2p));
 }
-
